Support running btests on Windows

timwoj opened the pull request:
This PR covers the majority of the changes that need to happen to the btest script in order to run tests on Windows. There's a few known issues I'll go into at the end. The big differences here are:

- Switching to the `multiprocess` module instead of the built-in `multiprocessing` module, since the default one has some problems with pickling our internal types.
- Changing executing processes on Windows to wrap them in a simple bash script, which lets them run correctly under msys2 (the bash environment installed with Git on Windows).
- Passing data from the test manager down into the child processes via a proxied dictionary. `mulitprocess` on Windows doesn't seem to like our use of globals here and complains that it can't find objects. The proxied dictionary works around this, and is probably the more correct way of doing this anyways.

Known issues:
- Running tests against debug builds hits an assertion down in the Windows OS libraries. I'm pretty sure this is a libkqueue bug, and opened https://github.com/mheily/libkqueue/issues/151 to have them investigate. The easy work-around is to not close the event_queue file descriptor in `iosource_mgr`, and that can be used for testing.
- Shutting down a run with the `-j` flag with ctrl-c doesn't work right. There's something different about signal masking on Windows. I spent a day trying to figure this one out and haven't solved it yet.
- There's a problem with pickling our internal types using the `dill` library used by `multiprocess` that changes that internal ID stored for the types. `run_cmdseq` previously used `isinstance` to check whether a command being run was a `CmdSeq` but because the type IDs are different it was returning false (on all platforms). There's currently a work-around where we just check the type name, but it's likely possible to fix this via implementing custom pickling methods for those types.

These changes will require some modifications to the Zeek tree to work right as well, but this PR is independent of those.

ckreibich commented on btest:
```diff
@@ -252,7 +252,17 @@ def runTestCommandLine(cmdline, measure_time, **kwargs):
+                with os.fdopen(tf, 'wb') as f:
+                    f.write(fcontents.encode('utf-8'))
+
+                cmd = ['bash.exe', '-c', normalize_path(tf_name)]
```
This codifies that we're using Bash on Windows. I'm fine with that if we can't do this any other way, but should we somehow test availability explicitly at startup? We can always revisit later, but it'd be good to understand why this is needed a little better. I dug through the subprocess docs re. Windows and couldn't find anything, sigh.

timwoj replied:
Added a check at startup to make sure bash.exe exists and is executable.

ckreibich reviewed (commented):
This must have been pretty darn painful. Just a few quick comments here as you're finalizing this.

timwoj commented on setup.py:
```diff
@@ -1,6 +1,6 @@
 #! /usr/bin/env python
 
-from setuptools import setup
+from distutils.core import setup
```
Huh, I don't remember what pycharm was complaining about, but after digging a bit more it appears that distutils was deprecated a long while back. I'll revert that.

timwoj commented on btest:
```diff
@@ -1845,6 +1845,16 @@ if __name__ == '__main__':
+        type="str",
+        default="build",
+        help=
+        "The relative path to the build directory, from the main Zeek repository. Defaults to 'build'."
```
How about an option that lets you set generic default option values?

ckreibich replied:
Ah, I get it now. Zeek's btest.cfg hardwires the build directory in a couple of places, and we currently don't provide a way to override this.

There's an existing feature that's pretty close: btest dumps the entire environment into its underlying `ConfigParser` instance defaults, and we use default interpolation. So if you set e.g environment variable `BUILD`, you can grab that in the btest.cfg via `%(BUILD)s`, including in the environment sections. But this substitution will fail loudly if you don't have `BUILD` in your environment, and that's arguably not great here — we'd still like it to default to `build`.

So yeah — generic default option values sound right to me. A way to get more `ConfigParser` defaults in place, something like `btest --set key=val`. (That would match what `zeek-client` does, but could be something else.)

ckreibich replied:
A thought: we could add a new section

```
[defaults]
BUILD=build
```

that would still be overriden by the environment, and make it `btest --set-default KEY=VAL` that'd have final say.

timwoj replied:
This sounds good to me.

timwoj replied:
Fixed in 4185d71

timwoj commented on btest:
```diff
@@ -1084,7 +1084,13 @@ class Test(object):
+                    # different type upon un-serializing, failing
+                    # isinstance(). So we take the class name as a sufficent
+                    # signal.
+                    if type(cmd).__name__ == 'CmdSeq':
```
Agreed. We can always open an issue against it and look into it later.

timwoj commented on btest:
```diff
@@ -837,6 +837,7 @@ class Test(object):
         (name, ext) = os.path.splitext(name)
 
         name = name.replace("/", ".")
+        name = name.replace("\\", ".")
```
Yeah, that would work.

timwoj commented on btest:
```diff
@@ -1065,7 +1065,12 @@ class Test(object):
+                if not self.mgr_data["Options"].tmps:
+                    # Close these here, since the rm command in rmtmp will return an error on
+                    # windows if the files are still open.
+                    self.log.close()
```
Sounds reasonable. I'll fix that.

bbannier commented on btest:
```diff
@@ -2134,21 +2134,26 @@ if __name__ == '__main__':
+    if sys.platform == 'win32':
+        # On win32 we have to use a named pipe so that python's multiprocessing
+        # chooses AF_PIPE as the family type.
+        addr = f'\\\\.\\pipe\\btest-pipe-{os.getpid()}'
```
Note that #81 has been merged, and you should if needed change the minimum version checked in CI. We should always check the documented minimal version.

That being said, with tools like [`pyupgrade`](https://pypi.org/project/pyupgrade/) I don't think we need to block this PR until that decision is made. Instead, let go of your dear f-strings for now and then after the decision for a target version has been made clean up the code automatically with e.g., `pyupgrade --py36-plus <SOURCE>...` (or whatever other minimal version we land on). With that we'd use f-strings in even more places 😉

timwoj replied:
Removed them for now.

awelzel commented on btest:
```diff
@@ -412,24 +479,33 @@ class TestManager(multiprocessing.managers.SyncManager):
+    # at run time. This basically segments the list of tests into chunks and runs
+    # until we're out of chunks.
+    def threadRun(self, thread_num, mgr_data):
+        # This should prevent the child processes from receiving SIGINT signals and
```
I think this is the main one I'd like to see considered, but potentially a bit controversial. I just find the `self.mgr_data` changes intrusive, noisy and  not better than the globals.

Instead of all the `self.mgr_data` lookup changes that follow, could we instead fixup the globals in the childs so they exist ? Can argue globals aren't nice and all, but they are there today and what from what I can tell are just read-only anyway.

Essentially, run the following here and then :crossed_fingers: that the rest of the changes are not needed anymore (assuming the keys stay consistent which I think isn't the case for the some).
```
     # Fixup globals in Windows sub processes
    for (global_key, global_value) in mgr_data.items():
        globals()[global_key] = global_value
```

I've tested a minimal example with multiprocess on Linux and that seems to work out and would be surprised why it wouldn't on Windows if all the `self.mgr_data[...]` lookups work out there.

timwoj replied:
Yep, that looks like it'll work on Windows too.

awelzel replied:
Thanks for giving that a shot.

awelzel replied:
Hmm, I'm a bit confused about the location where it's now in `Test.run()` - isn't it enough to fix them once when the process started?

awelzel commented on btest:
```diff
@@ -1266,40 +1363,51 @@ class Test(object):
-                subprocess.call("rm -rf %s 2>/dev/null" % self.tmpdir, shell=True)
+                if sys.platform == 'win32':
+                    win_path = pathlib.PureWindowsPath(self.tmpdir)
+                    subprocess.call(f"rmdir /s /q {str(win_path)}", shell=True)
```
This shouldn't need `shell=True` when using the args form:
```
subprocess.call(["rm", "/s", "/q", str(winpath)])
```

I think the linux one could also use stderr=subprocess.DEVNULL and could let go of `shell=True`.

timwoj replied:
The whole `win32` block can actually just go away, since we're running under bash. This works just fine under Windows as well:

```
subprocess.call("rm -rf %s 2>/dev/null" % self.tmpdir, stderr=subprocess.DEVNULL)
```

awelzel replied:
This currently looks a bit funky with the `redirect` and `stderr=DEVNULL` does the equivalent `call(["rm", "-rf", self.tmpdir], stderr=subprocess.DEVNULL)` work on windows without the shell=True part?

Otherwise, would be nice to drop the redirect from the commadline.
```
1384             if os.path.isdir(self.tmpdir):                                      
1385                 subprocess.call("rm -rf %s 2>/dev/null" % self.tmpdir,          
1386                                 stderr=subprocess.DEVNULL,                      
1387                                 shell=True)  
```

timwoj replied:
Surprisingly, the lack of a `shell=True` broke the build on ubuntu.

timwoj replied:
Removing the redirect works, but changing to the list for the commands breaks ubuntu for some reason. It doesn't remove the files, and then complains they still exist at the end of the run (https://github.com/zeek/btest/actions/runs/3850623526/jobs/6560971916, for example).

awelzel replied:
I'm not fully sure why it doesn't more loudly fail, but if you pass a list, need to remove `shell=True`.
```
In [19]: subprocess.call(["rm", "-rf", "/tmp/xyz"], shell=True)
rm: missing operand
Try 'rm --help' for more information.
```

```
In [20]: subprocess.call(["rm", "-rf", "/tmp/xyz"])
Out[20]: 0
```

Does a list without `shell=True` also work on windows?

timwoj replied:
All of the platforms seem happy with `subprocess.call(['rm', '-rf', self.tmpdir], stderr=subprocess.DEVNULL)`, so we'll go with that.

awelzel reviewed (changes requested):
I've made a bit of a pass.

awelzel commented:
> Switching to the multiprocess module instead of the built-in multiprocessing module, since the default one has some problems with pickling our internal types.

From the naive side: Is there something we fix/work-around in the types involved? Adding non-stdlib dependency when there wasn't one before seems unfortunate to me.

timwoj commented on btest:
```diff
@@ -761,7 +772,7 @@ class Test(object):
         self.stdout = None
         self.stderr = None
-        self.tmpdir = None
+        self.tmpdir = TmpDir
```
Fixed this one in a later commit.

timwoj commented on btest:
```diff
@@ -165,14 +186,32 @@ def ExpandBackticks(origvalue):
+        if tf_name:
+            try:
+                os.remove(tf_name)
+            except:
```
Agreed, it was a temp fix for something else. I'm pretty sure we only care about a FileNotFoundError here.

timwoj commented on btest:
```diff
@@ -968,10 +984,13 @@ class Test(object):
+        self.diag = normalize_path(os.path.join(self.tmpdir, ".diag"))
+        self.verbose = normalize_path(os.path.join(self.tmpdir, ".verbose"))
+        self.baselines = [
+            make_posix_path_from_windows(os.path.join(d, self.name))
```
I finally went back to look at this one, because one of the btest btests is failing due to this change. The reason we change the baseline directories specifically from "C:\..." paths to "/c/..." paths is because `btest-diff` does this:

```
IFS=':' read -ra baseline_dirs <<<"$TEST_BASELINE"
```

I could technically make that check for MINGW/Cygwin and look at ";" instead. It'd require one other change in `btest` as well, but it's minor.

timwoj commented:
I rebased and squashed down all of the fixup commits. I'm down to just two btests failing:

- tests.diag-file: This one looks like the .stdout and .stderr files get reset between commands run (meaning each time a bash script is fired up), which shouldn't be happening. This causes the output from one the internal btest commands to get lost, which the tests requires for success.
- tests.environment: This one fails because the test calls `pwd` in the middle of it and compares it to the `testbase` value read from btest.cfg. On Windows, `pwd` will return a POSIX-style path (so `/c/...`) whereas btest currently uses pseudo-Windows-style paths internally (`C:/...`). It's making me want to switch everything over to use POSIX-style paths since it would help with some consistency between POSIX and Windows in other places, but the last time I tried to do that everything went south quickly.

I'll look over the rest of the comments above after I get back from the holiday break.

timwoj commented on btest:
```diff
@@ -238,7 +277,17 @@ def runTestCommandLine(cmdline, measure_time, **kwargs):
+                tf, tf_name = tempfile.mkstemp(suffix='.sh', dir=tmpdir)
+                fcontents = f'#!/usr/bin/env bash\n{" ".join(args)}\n'
+                with os.fdopen(tf, 'wb') as f:
+                    f.write(fcontents.encode('utf-8'))
```
Looking at this again, it might actually be better to use `tempfile.TemporaryFile()` and just write to it directly. The file opened by `TemporaryFile` is also deleted automatically when the variable goes out of scope, so no need to clean it up manually.

timwoj commented on btest:
```diff
@@ -449,17 +525,17 @@ class TestManager(multiprocessing.managers.SyncManager):
                     continue
 
-                if t.serialize and hash(t.serialize) % Options.threads != thread_num:
+                if t.serialize and int(hashlib.md5(t.serialize.encode('utf-8')).hexdigest(), 16) % self._options.threads != thread_num:
```
> This was very strange to look at, because self.serialize = [] in Test.__init__() and a list is neither hashable nor md5'ble.
> 
> It's unrelated, but could you fix that constructor to set either None or "" rather than [] which is actively confusing?

Hm, good catch. I'd only ever seen single-string data in that, but apparently `mergePart` will append the serialize data to another test part, which might break with `None`. I'll look for a test that actually exercises that part of the code and see what happens with making it a string.

> Also, why the change from hash() to md5() ?

Yep, see the commit message on 0524d07cdeaffe0845ebc100c509a4e9c5abe08c:

> Windows has some issue where `hash()` returns different values for the
same string in the different child processes. md5() returns the same
values in each.

timwoj replied:
The serialize info comes from `@TEST-SERIALIZE` commands in a script. According to [the documentation](https://github.com/zeek/btest#id29), these values are always an "arbitrary user-chosen string". In the command parsing code, we set up a regex for those commands and set the `multiple` flag to false, which means it can only be set once and should only be a single string value. `Test.serialize` is then set to that string value, overwriting the default array value it's set to in the `Test` constructor. I think `Test.mergePart()` is expecting it to be an array when it does `self.serialize += part.serialize`, but that's clearly not the case. Otherwise, it would be just concatenating the strings.

timwoj commented on btest:
```diff
@@ -969,10 +1048,17 @@ class Test(object):
-        self.baselines = [os.path.abspath(os.path.join(d, self.name)) for d in BaselineDirs]
+        # Pull the proxied data from the manager and save it since it's going to be used
+        # more later
+        self.mgr_data = mgr_data
```
It moved in a later version.

timwoj commented on btest:
```diff
@@ -969,10 +1048,17 @@ class Test(object):
+        self.tmpdir = normalize_path(os.path.join(self.mgr_data["TmpDir"], self.name))
+        self.diag = normalize_path(os.path.join(self.tmpdir, ".diag"))
+        self.verbose = normalize_path(os.path.join(self.tmpdir, ".verbose"))
+        self.baselines = [
```
Seems reasonable to me.

timwoj commented on btest:
```diff
@@ -229,7 +268,7 @@ def runTestCommandLine(cmdline, measure_time, **kwargs):
         return Timer.timeSubprocess(cmdline, **kwargs)
     (success, rc) = runSubprocess(cmdline, **kwargs)
-    return (success, rc, -1)
+    return success, rc, -1
```
The parens are redundant since python will just return the tuple as-is without them. I'll put them back if the style matters though.

timwoj commented on btest:
```diff
@@ -1214,33 +1310,34 @@ class Test(object):
             (success, rc) = self.execute(filter, apply_alternative=None)
             if not success:
-                return (False, rc)
+                return False, rc
```
See earlier comment.

timwoj commented on btest:
```diff
@@ -4,20 +4,19 @@
-import multiprocessing
-import multiprocessing.managers
-import multiprocessing.sharedctypes
+import multiprocess as mp
```
Fixed. And no, this is the first non-stdlib requirement for btest.

timwoj commented:
> From the naive side: Is there something we fix/work-around in the types involved? Adding non-stdlib dependency when there wasn't one before seems unfortunate to me.

Prior to switching over to the other library, I did attempt to fix the types it was failing on. It turns into a game of whack-a-mole pretty fast, with one type after another failing for varying reasons. I got fairly far into it and then ran into a type that I couldn't figure out how to fix, and figured there had to be a better way to go about it than manually fixing every type and hoping I didn't miss something.

awelzel commented on btest:
```diff
@@ -323,18 +374,31 @@ class TestManager(multiprocessing.managers.SyncManager):
         self._num_tests = None
         self._timing = None
         self._ports = None
+        self._options = options
```
With the globals fixup, in the `Test` code we reference `Options`. Might be more consistent to revert the manager's usage of Options also back to the global.

(This makes the assumption you're okay with keeping the globals around, but seems that could work out).

timwoj replied:
I'm not the biggest fan of the globals but I agree that it minimizes the changes in this PR. We can always (hopefully?) revisit those later and see if we can find a better way to remove them.

timwoj replied:
This one continues to be very weird. I backed out the `_options` changes, and tests started failing because `Options` wasn't defined inside of `TestManager.threadRun`. That makes sense because `threadRun` is run inside another process. To fix that I moved resetting the globals up from `Test.run` into `threadRun` which should be fine since `threadRun` just calls `run` directly.

And now, `run` is complaining that it can't find the globals which clearly don't exist if I print out the result of `globals()`. What the heck, Python?

awelzel replied:
> EDIT: I can obviously reset the globals in both places, but that seems like should be unnecessary. The test I'm using to check this is tests.unstable which started failing after the changes. That test passes -z 4 to btest, forcing it to retry a few times. Everything looks fine on the first attempt, but things are broken on the second attempt.

Seems I can reproduce this on Unix with your branch by setting the start_method to "spawn".

```
Traceback (most recent call last):
  File "/home/awelzel/corelight-oss/btest/.venv/lib/python3.9/site-packages/multiprocess/process.py", line 315, in _bootstrap
    self.run()
  File "/home/awelzel/corelight-oss/btest/.venv/lib/python3.9/site-packages/multiprocess/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/awelzel/corelight-oss/btest/testing/../btest", line 456, in <lambda>
    t = mp.Process(name="#%d" % (i + 1), target=lambda: self.threadRun(i, mgr_data))
  File "/home/awelzel/corelight-oss/btest/testing/../btest", line 527, in threadRun
    t.run(self, mgr_data)
  File "/home/awelzel/corelight-oss/btest/btest", line 1077, in run
    self.tmpdir = normalize_path_join(TmpDir, self.name)
NameError: name 'TmpDir' is not defined
1 of 1 test failed
```

Let me stare a bit.

awelzel replied:
Hmm, hmm, I think I grasp what's going on.

The `Test` objects that are created during `def rerun()` use a class from the child process. The ones that are created initially and passed down are from a "different" class specific to the manager process. Methods/functions of these two use different `globals()`.

So, the following in `def clone()` and moving the fixup into `threadRun()` seems to work here.
```
@@ -976,7 +994,7 @@ class Test(object):
     # Copies all control information over to a new Test but replacing the test's
     # content with a new one.
     def clone(self, content=None, increment=True):
-        clone = Test("")
+        clone = self.__class__("")
         clone.number = self.number
         clone.basename = self.basename
         clone.name = self.basename
```

awelzel replied:
The windows-latest Python 3 CI pipeline looks okayish for this: There are some failures but seem unrelated.
https://github.com/zeek/btest/actions/runs/3854803832/jobs/6569182854

The Python 3.5 one fell apart completely though (it's green, but didn't actually run a test and failed with TmpDir directly). I've added Python 3.7 and that looks better.

timwoj replied:
Alright I think we managed to fix this one finally.

bbannier commented on setup.py:
```diff
@@ -43,4 +43,5 @@
         'Topic :: Utilities',
     ],
-)
+    install_requires=['multiprocess'],
+    python_requires='>=3.7')
```
I tried to take this for a spin, but something seems missing about installing this package with `pipx` (which should just use `pip` and `virtualenv`s under the covers):

```console
$ pipx install git+https://github.com/zeek/btest@topic/timw/windows-support-2 
creating virtual environment...
installing btest from spec 'git+https://github.com/zeek/btest@topic/timw/windows-support-2'...
done! ✨ 🌟 ✨
  installed package btest 0.72.post20, installed using Python 3.11.1
  These apps are now globally available
    - btest
    - btest-ask-update
    - btest-bg-run
    - btest-bg-run-helper
    - btest-bg-wait
    - btest-diff
    - btest-diff-rst
    - btest-progress
    - btest-rst-cmd
    - btest-rst-include
    - btest-rst-pipe
    - btest-setsid

$ btest -l
Traceback (most recent call last):
  File "/Users/bbannier/.local/bin/btest", line 34, in <module>
    import btest_lib.outputhandlers
ModuleNotFoundError: No module named 'btest_lib'
```

I don't know much about setuptools, so I added `packages=['btest_lib]` to `setup` (required dropping `package_dir` for testing). This still had problems since `multiprocess` wasn't properly installed.

timwoj replied:
Yep, I was working on that on Friday. It continues today.

timwoj replied:
I ended up just backing out the change that split the output handlers into a separate file. I spent a while trying to get the setup stuff working, didn't make much progress, and decided it wasn't worth the effort for something that isn't really necessary.

timwoj commented:
> tests.environment: This one fails because the test calls `pwd()` in the middle of it and compares it to the `testbase` value read from btest.cfg. On Windows, `pwd` will return a POSIX-style path (so `/c/...`) whereas btest currently uses pseudo-Windows-style paths internally (`C:/...`). It's making me want to switch everything over to use POSIX-style paths since it would help with some consistency between POSIX and Windows in other places, but the last time I tried to do that everything went south quickly.

I'm not entirely sure how to fix this one. I tested changing `normalize_path` to return POSIX-style paths, and all that leads to is basically every os-level call (`open`, `os.chdir`, etc) because the paths don't exist. That's probably what I saw the first time around as well.

timwoj commented:
All of the btest btests finally pass. This is ready for another review pass.

timwoj commented on btest-diff:
```diff
@@ -225,7 +225,7 @@ if [ -n "$baseline" ]; then
             diff -s "$@" "$canon_baseline" "$canon_output" >>$TEST_DIAGNOSTICS
         else
-            diff -au "$@" "$canon_baseline" "$canon_output" >>$TEST_DIAGNOSTICS
+            diff -au --strip-trailing-cr "$@" "$canon_baseline" "$canon_output" >>$TEST_DIAGNOSTICS
```
This change required updating the Alpine Dockerfile in the main Zeek repo to install a newer version of `diff`. That was the only platform that had any problem with it.

awelzel reviewed (commented):
One more set of comments.

timwoj commented on testing/tests/env-var-casing.test:
```diff
@@ -1,5 +1,6 @@
-# %TEST-DOC: Validates that env vars are case-sensitive; this is a regression test for #75.
+# %TEST-DOC: Validates that env vars are case-sensitive; this is a regression test for #75. Environment variables on Windows are always uppercase, due to legacy DOS requirements. This test will be skipped on that platform.
 #
+# %TEST-REQUIRES: "$(uname -s | cut -c 1-5)" == "MINGW"
```
This should probably change to use the `is-windows` script. I'll look at why I was skipping this test in the first place.

timwoj replied:
Ah, there's a big comment there why. I switched it over to be `! ${SCRIPTS}/is-windows`.

timwoj commented on btest:
```diff
@@ -46,9 +47,25 @@ except KeyError:
 def output(msg, nl=True, file=None):
     if not file:
-        file = sys.stderr
+        file = open(sys.__stderr__.fileno(),
```
It reopens stdout/stderr, but with `\n` as the line ending. Otherwise writing to those on Windows gets `\r\n` which breaks a lot of things.

timwoj commented on btest:
```diff
@@ -158,21 +175,55 @@ def readStateFile():
+        if sys.platform == 'win32':
+            try:
+                tf, bash_cmd = _build_win_subprocess_cmd_script(cmd, None)
+                pp = subprocess.Popen(cmd, stdout=subprocess.PIPE)
```
Odd, because it's definitely functioning or none of the zeek btests would work. We expand zeek-path-dev into the PATH with that.

timwoj commented on btest:
```diff
@@ -755,6 +859,12 @@ class Test(object):
+        if not self.serialize:
+            return None
+
+        return int(hashlib.md5(self.serialize.encode('utf-8')).hexdigest(), 16)
```
Looks like that works just fine. I switched over to it.

ckreibich commented on testing/tests/multiple-baseline-dirs.test:
```diff
@@ -41,5 +41,5 @@
 %TEST-START-FILE btest.cfg
 [btest]
-BaselineDir = baseline1:baseline2:baseline3
+BaselineDir = baseline1%(pathsep)sbaseline2%(pathsep)sbaseline3
```
I wonder whether we should warn about non-portable aspects btest encounters in tests or test configs. Most people will never notice that using ":" breaks on Windows, but btest could flag it. For example, the `PATH` specified in Zeek's btest.cfg uses `:`. (For a future PR.)

One way or another we should cover this in the README.

timwoj replied:
Part of the problem is that you're going to get something like `C:/some/path:C:/some/other/path`. How do you check to see that the colons there are part of the path or that they're being used as path separators?

I agree that there should probably be a "Windows Caveats" section added to the readme.

timwoj replied:
Added a Windows Caveat section to the README.

ckreibich commented on btest:
```diff
@@ -274,7 +274,11 @@ def runSubprocess(*args, **kwargs):
                 tf, tf_name = tempfile.mkstemp(suffix='.sh', dir=tmpdir)
-                fcontents = '#!/usr/bin/env bash\n%s\n' % (" ".join(args))
+                if len(args) > 1:
+                    cmd = shlex.join(args)
```
Why shlex? For added escaping? Could use a comment.

timwoj replied:
See https://github.com/zeek/btest/pull/80#discussion_r1049554081

ckreibich replied:
Got it, resolved. Thanks!

ckreibich commented on btest:
```diff
@@ -203,11 +221,8 @@ def ExpandBackticks(origvalue):
-                os.remove(tf_name)
-            except FileNotFoundError:
-                pass
+        if tf:
```
Looks like you could use `with tf:` above too, as below. Or, if I'm missing something, perhaps a `try`/`finally`.

timwoj replied:
It's like this because the file should exist at least through the call to `pp.communicate`. We don't want to wrap that in a `with` because `tf` doesn't exist in the non-Windows case. Alternatively, I could just duplicate the `communicate` and `decode` lines in both blocks and wrap it all in a `with` in the Windows one.

ckreibich reviewed (commented):
Thanks a bunch for keeping the individual commits so manageable. It's very close for me — the biggie is likely that I'm wondering whether we actually need to use `multiprocess` on non-Windows, or could continue to use `multiprocessing` there.

I think it'd be fair to replace Robin with `The Zeek Team` in `setup.py` at this point. :-)

awelzel commented on btest:
```diff
@@ -33,6 +30,10 @@ import xml.dom.minidom
 
 from datetime import datetime
 
+import multiprocess as mp
```
> Do you think we could get by with only needing multiprocess on Windows, and sticking with multiprocessing elsewhere? It only really matters on Windows and the usage looks like a drop-in replacement, no?

This is a great thought/idea!

timwoj replied:
> Do you think we could get by with only needing multiprocess on Windows, and sticking with multiprocessing elsewhere? It only really matters on Windows and the usage looks like a drop-in replacement, no?

I don't see why not. I'll put it in and see what CI thinks about it.

timwoj replied:
CI seems content. Resolving.

timwoj commented on testing/Scripts/strip-test-base:
```diff
@@ -3,4 +3,10 @@
 dir=$(dirname "$0")
 testbase=$(cd "$dir/.." && pwd)
+
+if [ "$(uname -s | cut -c 1-5)" == "MINGW" ]; then
```
Replaced with a call to `${SCRIPTS}/is-windows`

timwoj commented on btest:
```diff
@@ -175,23 +175,41 @@ def readStateFile():
+                                     suffix='.sh',
+                                     dir=tmpdir,
+                                     delete=True)
+    fcontents = '#!/usr/bin/env bash\n%s\n' % cmd
```
I would hope so, in that this is running under `bash` as it is.

timwoj commented on README:
```diff
@@ -330,6 +330,12 @@ and 1 otherwise. Exit code 1 can also result in case of other errors.
     Markdown. In the output each test includes the documentation
     string that's defined for it through ``@TEST-DOC``.
 
+-s key=value, --set=key=value
```
Added a test and example, but I somehow broke the table of command-line arguments in the README. I'll look at it more closely once I have everything else resolved.

timwoj commented on .github/workflows/btest.yml:
```diff
@@ -9,23 +9,44 @@ jobs:
-        # it.
-        python-version: [3, 3.5]
-        os: [macos-latest, ubuntu-20.04]
+        python-version: [3, 3.7]
```
How are we doing it with other projects?

timwoj replied:
Looks like zeek-client is doing `["3.7", "3.8", "3.9", "3.10", "3.11"]`. I'm good with that.

ckreibich replied:
I learned yesterday that you can do ranges in those things, and apparently something like `'3.7 - 3` might work. I haven't tried. It'd have the benefit of automatically pulling in new 3.x releases. No biggie.

timwoj commented on btest:
```diff
@@ -47,6 +48,12 @@ except KeyError:
 
+def normalize_path(path):
+    os_path = os.path.abspath(path)
+    windows_path = pathlib.PureWindowsPath(os_path)
```
It's about doing the right thing on Windows, mostly to convert them from C:\path to C:/path so bash works correctly.

timwoj replied:
Add a comment to the function about what it's doing.

ckreibich reviewed (approved):
LGTM! :-)
